Trilinos: Added Cuda support

amklinv-nnl opened the pull request:
@aprokop @keitat I would like to iterate on this PR.  I used it to "successfully" install Trilinos, but no tests were run, so this is not ready to be merged in its current state.

amklinv-nnl committed:
Trilinos: Added Cuda support

aprokop commented:
@amklinv-nnl Thanks for doing this! It's a dirty work, but needs to be done.

I think this may be the right first step, but more work is likely needed. I think we would want to look closer at the kokkos package in spack, and align this with it. Right now, I see at least few things to look at:
- should this also introduce `+openmp` variant, or is that left for later?
- it should allow to specify Cuda architecture; this will likely mean a lot of copy paste
- depending on `kokkos-nvcc-wrapper` is tricky; it should be done when using Nvidia compiler, but not with Clang (newer versions support Cuda)
- how do we deal with Cuda version dependency? Do we know which versions of Trilinos depend on which minimum version of Cuda? (probably not a big deal, because even kokkos does not specify that)

amklinv-nnl commented:
- [x] Introduce OpenMP variant
  - already existed
- [ ] Allow the user to specify Cuda architecture (as in Kokkos package)
- [ ] Allow the user to use nvcc_wrapper or clang
- [x] Add testing support to Trilinos package
- [ ] Fix Epetra/Tpetra/Xpetra bug (which can trigger when someone tries to ```spack install trilinos``` without modifying the variants)

amklinv-nnl commented:
Ran into an error unrelated to Cuda, but might as well fix it while I'm here:
```
  >> 979    CMake Error at packages/xpetra/CMakeLists.txt:120 (MESSAGE):
     980      If Xpetra Epetra support is enabled and Epetra 32-bit global indices are
     981      enabled, Xpetra requires that you enable Tpetra_INST_INT_INT
```

jrood-nrel commented:
Would love to see this finally get in and am open to help. I have a custom Trilinos with a CUDA variant I've been using here https://github.com/jrood-nrel/spack-configs/blob/master/custom/packages/trilinos/package.py .

amklinv-nnl commented:
I appreciate it!  I wrote a similar custom Trilinos package a while back that looked very similar to yours, but spack recently(?) introduced a kokkos_nvcc_wrapper package that solves the MPI problem for us.

amklinv-nnl committed:
Trilinos: Add the ability to run tests

amklinv-nnl commented:
So...did anybody else notice that there are conflicts that refer to the xpetra variant, but there is no xpetra variant?  Should there be an xpetra variant?

aprokop commented:
> So...did anybody else notice that there are conflicts that refer to the xpetra variant, but there is no xpetra variant? Should there be an xpetra variant?

The only thing that I'm aware of is that in the past year or so MueLu changed its behavior to deal with 32-bit indices, and this must have affected xpetra. It must depend on the Trilinos version, too.

jrood-nrel commented:
Maybe we can add `generator = 'Ninja'` while we're at it?

amklinv-nnl commented:
> Maybe we can add `generator = 'Ninja'` while we're at it?

Happy to do so.  Can you explain why you would like this?

jrood-nrel commented:
So I actually tried using Ninja recently with Trilinos and the Fortran/C mixed language verification failed during CMake configure, even though Ninja 1.10.0 should now support Fortran. So I'm a little confused about that since I've used it in Trilnos in the past. Ninja is a build system which is more efficient than the default Unix makefiles CMake generates. If you build Trilinos with Ninja makefiles generated from CMake, you can save a significant amount of build time. I recommend doing so to see the performance advantage if you haven't.

sethrj commented:
@jrood-nrel since "make" is only run once during a standard spack install, Ninja doesn't buy you anything. In fact, it costs more for two ways:
1. It requires another dependency (and possibly its dependencies as well), and this also means keeping multiple installations of ninja for each compiler toolchain (ugh) since spack can't currently share build-only dependencies across compilers
2. CMake's Ninja generator for some reason takes a substantial amount of time  during the "generate"  phase if you have a lot of libraries linked together. Trilinos, because of TriBITS, links **everything**  together, so for some trilinos/tribits projects I've seen Ninja take ~5 minutes during the "generating"  phase.

I haven't seen any  other spack packages enable ninja builds, so I'd recommend leaving it at the default for this one.

jrood-nrel commented:
I disagree @sethrj . I guess this probably highly depends on each person's situation, but I think over time, everyone would benefit from having Ninja enabled for Trilinos. Ninja only depends on Python and it's likely a person either has Python set as an external or already built, and Ninja takes an insignificant amount of time to build in my opinion. Personally I build Trilinos a _lot_, like every day, and on a machine with an SSD, I would much rather have Ninja enabled by default. I'm guessing I'm not the only who uses Spack who has built Trilinos more than once for things they work on. We see significant performance benefits when building Trilinos with Ninja using our own scripts. This is how long it just took me to build Ninja:
```
ninja: Successfully installed ninja
  Fetch: 0.01s.  Build: 28.06s.  Total: 28.07s.
```

Edit: To give some more perspective, we get a 20-30% speedup using Ninja in Trilinos.

jrood-nrel commented:
Here are some other packages using Ninja as a generator:
```
aluminum/package.py:    generator = 'Ninja'
archer/package.py:    generator = 'Ninja'
dbcsr/package.py:    generator = 'Ninja'
dihydrogen/package.py:    generator = 'Ninja'
fairlogger/package.py:    # generator = 'Ninja'
gplates/package.py:    generator = 'Ninja'
hydrogen/package.py:    generator = 'Ninja'
lbann/package.py:    generator = 'Ninja'
llvm-openmp-ompt/package.py:    generator = 'Ninja'
py-torch/package.py:    # Use Ninja generator to speed up build times
```

amklinv-nnl committed:
Trilinos: add logic to enabling GO INT

This avoids the error message:
If Xpetra Epetra support is enabled and Epetra 32-bit global indices are
enabled, Xpetra requires that you enable Tpetra_INST_INT_INT.

amklinv-nnl committed:
Trilinos: Make long_long default GO type

Trilinos is not tested with GO long, so it should not be the default
ordinal type. Also added a conflict related to a MueLu bug which
occurs when building tests with GO long. This conflict should only
trigger if run_tests is enabled, but spack does not currently support
that logic in a conflict.

amklinv-nnl committed:
Trilinos: added xpetra variant

Some conflicts referred to a nonexistant xpetra variant.

psakievich commented on var/spack/repos/builtin/packages/trilinos/package.py:
```diff
@@ -337,6 +342,7 @@ class Trilinos(CMakePackage):
     depends_on('metis@5:', when='+metis')
     depends_on('suite-sparse', when='+suite-sparse')
     depends_on('zlib', when="+zlib")
+    depends_on('kokkos-nvcc-wrapper', when='+cuda')
```
@aprokop Does this work?  I've been playing around with this locally and I'm not getting `nvcc-wrapper` set as my compiler. Seems like I still need to explicitly set `nvcc_wrapper` as the `CXX` compiler in the package.

aprokop replied:
@psakievich I think you are right. It seems that one still has to explicitly set it as a compiler, too, using
```
options.append('-DCMAKE_CXX_COMPILER=%s' % self.spec['kokkos-nvcc-wrapper'].kokkos_cxx
```

amklinv-nnl replied:
Are you using MPI? kokkos-nvcc-wrapper should set environment variables that control which underlying compiler is picked up by MPI.

psakievich replied:
@amklinv-nnl I'm using MPI.  I was using inheritance to reconstruct this PR in a repo though so I could have missed something.  I'll take another look.

sethrj commented on var/spack/repos/builtin/packages/trilinos/package.py:
```diff
@@ -410,7 +416,8 @@ def define_tpl_enable(cmake_var, spec_var=None):
             define('Trilinos_VERBOSE_CONFIGURE', False),
-            define('Trilinos_ENABLE_TESTS', False),
+            define('Trilinos_ENABLE_TESTS', (
+                True if self.run_tests else False)),
```
```suggestion
                bool(self.run_tests)),
```

amklinv-nnl replied:
#19119 I would like to see run_tests actually do something for Trilinos, but if that's outside the scope of your PR, that's okay.

sethrj commented on var/spack/repos/builtin/packages/trilinos/package.py:
```diff
@@ -438,6 +445,17 @@ def define_tpl_enable(cmake_var, spec_var=None):
+        options.append(define_tpl_enable('CUDA'))
+        if '+cuda' in spec:
+            cxx_flags.extend([
+                '--expt-extended-lambda'
```
This is a NVCC flag: does it belong in the cxx_flags, or is that necessary due to the kokkos compiler wrapper?

sethrj commented on var/spack/repos/builtin/packages/trilinos/package.py:
```diff
@@ -715,7 +733,8 @@ def define_tpl_enable(cmake_var, spec_var=None):
                 define('Tpetra_INST_DOUBLE', True),
-                define('Tpetra_INST_INT_INT', gotype == 'int'),
+                define('Tpetra_INST_INT_INT', gotype == 'int' or
+                      '+epetra' in spec and '+xpetra' in spec),
```
I'm not sure Trilinos 13 allows multiple instantiations per build. Should this be a conflict? If left as-is, I'd add a set of parenthesis around the `and` fo improved readability.

psakievich commented:
@amklinv-nnl I tried this today and I'm not getting past cmake. Here are the files from my build system.  Looks like nvcc_wrapper is not getting called as the compiler.

[spack-build-out.txt](https://github.com/spack/spack/files/5293897/spack-build-out.txt). 
[CMakeError.log](https://github.com/spack/spack/files/5293895/CMakeError.log)
[CMakeOutput.log](https://github.com/spack/spack/files/5293896/CMakeOutput.log)

amklinv-nnl commented:
#19119 seems closer to completion, so I'm okay with closing this PR if you guys are.

tldahlgren commented:
Should this PR be closed since #19119 has been merged?

sethrj commented:
Thanks @amklinv-nnl ! This PR was foundational for getting the other CUDA PR merged -- I think the hangup for this one was trying to get trilinos tests compiled with arbitrary combinations of tribits packages 😂 😭
